CMakeLists.txt

```diff
@@ -115,6 +117,34 @@ if (NOT DEFINED CMAKE_CXX_EXTENSIONS)
     set(CMAKE_CXX_EXTENSIONS FALSE)
 endif()
 
+# Set the default warning level we use to pass the GitHub workflows
+if (QL_ENABLE_DEFAULT_WARNING_LEVEL)
+    if (MSVC)
+        # warning level 3
+        # There are also specific warnings disabled for MSCV in cmake/Platform.cmake.
+        add_compile_options(-W3)
+    else()
+        # lots of warnings
+        add_compile_options(-Wall -Wno-unknown-pragmas)
+    endif()
+endif()
+
+# Treat warnings on compile as errors
+if (QL_COMPILE_WARNING_AS_ERROR)
+    # all compiler warnings as errors
+    if (CMAKE_VERSION VERSION_GREATER_EQUAL "3.24.0")
+        # since v3.24 cmake can set all compiler warnings as errors itself
+        set(CMAKE_COMPILE_WARNING_AS_ERROR ON)
+    else()
+        # or set them manually
+        if (MSVC)
+            add_compile_options(-WX)
+        else()
+            add_compile_options(-Werror)
+        endif()
+    endif()
+endif()
+
```

Comment:
I think these can go into presets instead.

Reply:
In general presets are the right place for these default flags, but there is a couple of considerations:

- CMake version 3.19 and above is required for presets whereas `CMakeLists.txt` currently requires only 3.15.0. Since the goal of this change is to provide consistent default warnings everywhere, we should probably bump the minimum version to 3.19 if the defaults are in the presets.
- The presets in `CMakePresets.json` probably won't work for everyone out of the box, mainly because they are missing `BOOST_ROOT`, and there's obviously no way to know where all users have installed Boost. This means that the presets provided in `CMakePresets.json` are just examples for users to use (i.e. to make their own `CMakeUserPresets.json` file) but that brings us back to square one where users need to a) know where to look, and b) manually configure their build environment.

In this case I think the best solution might be to put the default compiler options in `Platform.cmake`, which is kind of in-between `CMakeLists.txt` and `CMakePresets.json`.

Reply:
One note, though: for me the goal of this would be to make it more convenient for contributors to use the same flags we have in the CI builds.  It's not to make those flags the default for users of the library.

Reply:
I don't mean to sound pedantic, but I think that the goal is both convenience as well as consistency. As pointed out by Ralf in the issue, there are already some default flags such as `/W3` in `QuantLib.vcxproj`, so it's a bit inconsistent that there aren't those same defaults in the CMake build settings too.

Also, again not trying to be pedantic, but since the CI build compiler flags are already in the presets, adding another preset won't provide contributions with any new information - they can already find the CI flags if they know where to look, which is the problem.

Ultimately it's your decision, but I still believe that harmonizing the default compiler flags across all supported build systems, along with a way to easily opt out of the defaults like Ralf has proposed, will provide the greatest benefit overall.

Reply:
>I don't mean to sound pedantic

So we skip `add_compile_options(-Wpedantic)`?! 😄

But seriously (and probably not a big surprise): I share Jonathan's opinion. The library is released warning free based on the levels defined in the ci and now this PR. If users extend the QL they now might get warnings compliant with contribution rules.

Luigi, are you afraid that users "run away" and stop using QL ones they see warnings? My experience over the years is that there are two kinds of developer: 95% don't care about warnings at all and 5% trying to avoid and fix them. But I haven't seen a single one who "ran away". 🏃‍♂️🏃‍♀️👟 And if they have other warnings defined they now can still opt-out.

Jonathan: Regarding `cmake` and _bump the minimum version to 3.19_: I would be careful here. `Ubuntu 20.04 LTS` default version is only [3.16.3](https://packages.ubuntu.com/search?keywords=cmake) (the corresponding github runner however is up-to-date). Not sure about older versions of VS prior to VS22 but I don't expect them to be up-to-date.

Reply:
By the way, how long are you planning to support all three build systems - CMake, Automake, and MSBuild?

In my experience a lot of projects are consolidating on CMake, so perhaps we should think about doing that too at some point? This would help address at least two of the concerns raised in Ralf's issue: the inconsistencies between different build system configs, and the learning curve for new contributors.

CMake 3.15 is packaged with Visual Studio 2019 so by the time that becomes the oldest supported version for QuantLib then perhaps we can drop support for Automake and MSBuild. What do you think?

If so, then maybe we can start thinking about deprecating Automake and MSBuild now to give users plenty of time to migrate.

Reply:
Discussing the deprecation of the current build systems is a little of-topic in a code review and definitely worth its on issue and/or the user group mail.

However, just a few words about it: Personally I'm in favor of `cmake`, especially for new projects.

But at one of my clients we have two big projects where `QuantLib.vcxproj` is only one part of a bigger VS solution of other `c++`- (incl. a legacy `COM`- and a `swig`-api) for the quant-side and `c#`-projects for database-access, GUI etc. It would be quiet some (unwelcome aka not budgeted) effort to migrate over from a pure `msbuild` to `cmake` included in a `msbuild`-environment. Especially as most of the colleagues probably never heard about `cmake` before. Currently we "just" replace one QL release by a newer one, run our tests and hope for the best.

I guess there other users might be in a similar situation.

Also for newbies like I was a couple of years ago the `QuantLib.sln` on Windows is still the easiest way for a plug-and-play start to try out QL?!

Reply:
> This said, this is definitely not the hill I want to die on 😄

Luigi, it is probably my poor English but I'm not sure what exactly the conclusion of your comment is: 🤔
- We can go with this PR but do not use `QL_ENABLE_DEFAULT_WARNING_LEVEL=ON` by default or
- you are not fully convinced but would give it a try?

Regarding where to put the above lines: I put them on purpose into `CMakeLists.txt` as it makes them more prominent and easier to find. Putting specific compile definition like `_SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING` into `Platform.cmake` is the purpose of that file but perhaps not global defaults like warning levels.

Or at least I would put them in an separate file like `./cmake/SetDefaultWarningLevels.cmake` and `include(SetDefaultWarningLevels)` here.

Reply:
It means I won't fight too much over this issue 😄 

I wouldn't use the strict warning level as a default for users' builds, and I'd use the compiler default instead. The compiler default was probably chosen for a reason — my guess is that people will ignore warnings if there is too many of them, so it's better to show them only important ones.

And I think that anybody that cares about this also knows enough about cmake to use presets and uses a recent version, so I'd put the settings there and be done with it.

This said, if I'm outvoted I'll merge this.

Reply:
> This said, if I'm outvoted I'll merge this.

Like this? :smirk:
- [x] `option(QL_ENABLE_DEFAULT_WARNING_LEVEL "Enable the default warning level to pass the ci pipeline" ON)`
- [ ] `option(QL_ENABLE_DEFAULT_WARNING_LEVEL "Enable the default warning level to pass the ci pipeline" OFF)`

Reply:
Perhaps one other thing we should consider is adding a [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) to tell new contributors what they need to know before they submit their PR. For example:

_The CI requires all changes to compile without warnings at the `-Wall` level on Linux and MacOS and the `/W3` level on Windows. Please make sure your pull request satisfies this requirement before submitting. If you have access to CMake you can set the `QL_ENABLE_DEFAULT_WARNING_LEVEL` CMake option to true to use the same compiler flags as the CI builds._

Reply:
True!

Especially as we have so many combinations of compiler versions in `linux-full-tests.yml` and `linux-nondefault.yml` that you hardly test them all "at home". 🏡